sapicore/engine/component/__init__.py

```diff
@@ -0,0 +1,183 @@
+
+        Parameters
+        ----------
+        data: Tensor
+            Input to be processed (e.g., added to a neuron's numeric state tensor `voltage`).
+
+        Returns
+        -------
+        dict
+            A dictionary whose keys are loggable attributes and whose values are their states as of this time step.
+            For potential use by a :class:`~pipeline.simulation.Simulator` or any other :class:`~pipeline.Pipeline`
+            script handling runtime operations.
+
+        Raises
+        ------
+        NotImplementedError
+            The forward method must be implemented by each derived class.
+
+        """
+        raise NotImplementedError
+
+    def heterogenize(self, unravel: bool = True):
+        """Edits configurable tensor attributes based on a sweep search dictionary if one was provided by the user
+        within this object's `configuration` dictionary.
+
+        Note
+        ----
+        If a `sweep` key is not present in `configuration`, this method will pass silently, retaining the existing
+        configurable values. It can be invoked safely from generic methods (e.g.,
+        :meth:`~engine.network.Network.build`).
+
+        Parameters
+        ----------
+        unravel: bool
+            When the property is a 2D tensor, dictates whether combination values should be assigned
+            to individual elements (True) or overwrite entire rows (False).
+
+            The former behavior is appropriate for synapse attributes. It is achieved using `numpy.unravel_index`,
+            which maps flattened-like indices to their nD equivalents given the shape of the array.
+
+            The latter behavior is called for when values are themselves 1D vectors (e.g., oscillator frequencies).
+
+        Warning
+        -------
+        This method mutates the underlying component object. To simply view the combinations, initialize a
+        :class:`~utils.sweep.Sweep` object with your desired search space dictionary and number of combinations.
+
+        """
+        # safe to call if object has no sweep specification or empty configuration; method will return silently.
+        if self.configuration and self.configuration.get("model", {}).get("sweep"):
+            # the sweep object maintains a list of valid parameter combinations, each of which is a dictionary.
+            sweep = Sweep(
+                search_space=self.configuration.get("model", {}).get("sweep"), num_combinations=self.num_units
+            )
+            # modify this instance in-place.
+            sweep.heterogenize(obj=self, unravel=unravel)
+
+    def state(self) -> dict:
+        """Returns a dictionary of this object's loggable properties and their states as of this simulation step.
+
+        Returns
+        -------
+        dict
+            Dictionary containing loggable property names and their values.
+
+        """
+        return {prop: getattr(self, prop) for prop in self._loggable_props_}
```

Comment:
This helped me spot what seems to be a bug / unintended behavior of the `self._loggable_props` cache, because of which applying your suggestion would sometimes lead to an exception (NoneType is not iterable). 

For configurables, after the attributes are set for the first time (see above), `self._config_props` seems to consistently find them. For loggables, however, `self._loggable_props` fills up only **after one of the loggable attributes is first accessed**. 

To replicate, run simulation.py in debug mode with a breakpoint at component/__init__.py, line 66 (the loggable initialization for loop). Run the loop in the debug console and note that `_loggable_props_` is full but `_loggable_props` is None. Then, type in `self.voltage` (or access any other loggable attribute in a similar way). Lo and behold, the `_loggable_props`  cache is full. 

This should be fixed when you get a chance to overhaul loggable. For now, though, I'd like to retain my workaround (dealing directly with _loggable_props_).